lib/Query/QueryBuilderConverter.php

```diff
@@ -0,0 +1,256 @@
+use Sulu\Component\DocumentManager\PropertyEncoder;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * Class which converts a Builder tree to a PHPCR Query.
+ */
+class QueryBuilderConverter extends ConverterPhpcr
+{
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+    /**
+     * @var Metadata[]
+     */
+    protected $documentMetadata = [];
+
+    /**
+     * @param PropertyEncoder
+     */
+    protected $encoder;
+
+    /**
+     * @param DocumentStrategyInterface
+     */
+    private $strategy;
+
+    /**
+     * @param SessionInterface $session
+     * @param EventDispatcherInterface $eventDispatcher
+     * @param MetadataFactoryInterface $metadataFactory
+     * @param PropertyEncoder $encoder
+     */
+    public function __construct(
+        SessionInterface $session,
+        EventDispatcherInterface $eventDispatcher,
+        MetadataFactoryInterface $metadataFactory,
+        PropertyEncoder $encoder,
+        DocumentStrategyInterface $strategy
+    ) {
+        $this->eventDispatcher = $eventDispatcher;
+        $this->qomf = $session->getWorkspace()->getQueryManager()->getQOMFactory();
+        $this->metadataFactory = $metadataFactory;
+        $this->encoder = $encoder;
+        $this->strategy = $strategy;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getQuery(QueryBuilder $builder)
+    {
+        $this->documentMetadata = [];
+        $this->sourceDocumentNodes = [];
+        $this->constraint = null;
+
+        $this->locale = $builder->getLocale();
+
+        if (!$this->locale) {
+            throw new \InvalidArgumentException(sprintf(
+                'No locale specified'
+            ));
+        }
+
+        $from = $builder->getChildrenOfType(
+            QBConstants::NT_FROM
+        );
+
+        if (!$from) {
+            throw new RuntimeException(
+                'No From (source) node in query'
+            );
+        }
+
+        $dispatches = [
+            QBConstants::NT_FROM,
+            QBConstants::NT_SELECT,
+            QBConstants::NT_WHERE,
+            QBConstants::NT_ORDER_BY,
+        ];
+
+        foreach ($dispatches as $dispatchType) {
+            $this->dispatchMany($builder->getChildrenOfType($dispatchType));
+        }
+
+        if (count($this->sourceDocumentNodes) > 1 && null === $builder->getPrimaryAlias()) {
+            throw new \InvalidArgumentException(
+                'You must specify a primary alias when selecting from multiple document sources ' .
+                'e.g. $qb->from(\'a\') ...'
+            );
+        }
+
+        $this->applySourceConstraints($builder);
+
+        $phpcrQuery = $this->qomf->createQuery(
+            $this->from,
+            $this->constraint,
+            $this->orderings,
+            $this->columns
+        );
+
+        $query = new Query($phpcrQuery, $this->eventDispatcher, $this->locale, [], $builder->getPrimaryAlias());
```

Comment:
I don't like the fact that we have to pass the `EventDispatcher` here... But that's out of scope of this PR, and probably not possible in a different way with the current architecture.